presto-main/src/main/java/com/facebook/presto/operator/SpillableGroupedTopNBuilder.java

```diff
@@ -0,0 +1,331 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.operator.Operator.NOT_BLOCKED;
+import static com.facebook.presto.operator.SpillingUtils.checkSpillSucceeded;
+import static com.facebook.presto.operator.WorkProcessor.TransformationState.finished;
+import static com.facebook.presto.operator.WorkProcessor.TransformationState.needsMoreData;
+import static com.facebook.presto.operator.WorkProcessor.TransformationState.ofResult;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+
+public class SpillableGroupedTopNBuilder
+        implements GroupedTopNBuilder
+{
+    private static final long INSTANCE_SIZE = ClassLayout.parseClass(SpillableGroupedTopNBuilder.class).instanceSize();
+
+    private final Supplier<InMemoryGroupedTopNBuilder> inputInMemoryGroupedTopNBuilderSupplier;
+    private final Supplier<InMemoryGroupedTopNBuilder> outputInMemoryGroupedTopNBuilderSupplier;
+    private final SpillerFactory spillerFactory;
+    private final List<Type> sourceTypes;
+    private final List<Type> partitionTypes;
+    private final List<Integer> partitionChannels;
+
+    private InMemoryGroupedTopNBuilder inputInMemoryGroupedTopNBuilder;
+    private InMemoryGroupedTopNBuilder outputInMemoryGroupedTopNBuilder;
+
+    private final LocalMemoryContext localUserMemoryContext;
+    private final LocalMemoryContext localRevocableMemoryContext;
+    private final AggregatedMemoryContext aggregatedMemoryContextForMerge;
+    private final AggregatedMemoryContext aggregatedMemoryContextForSpill;
+    private final DriverYieldSignal driverYieldSignal;
+    private final SpillContext spillContext;
+
+    private final long unspillMemoryLimit;
+
+    private Optional<Spiller> spiller = Optional.empty();
+    private ListenableFuture<?> spillInProgress = immediateFuture(null);
+
+    public SpillableGroupedTopNBuilder(
+            List<Type> sourceTypes,
+            List<Type> partitionTypes,
+            List<Integer> partitionChannels,
+            Supplier<InMemoryGroupedTopNBuilder> inputInMemoryGroupedTopNBuilderSupplier,
+            Supplier<InMemoryGroupedTopNBuilder> outputInMemoryGroupedTopNBuilderSupplier,
+            long unspillMemoryLimit,
+            LocalMemoryContext localUserMemoryContext,
+            LocalMemoryContext localRevocableMemoryContext,
+            AggregatedMemoryContext aggregatedMemoryContextForMerge,
+            AggregatedMemoryContext aggregatedMemoryContextForSpill,
+            SpillContext spillContext,
+            DriverYieldSignal driverYieldSignal,
+            SpillerFactory spillerFactory)
+    {
+        this.inputInMemoryGroupedTopNBuilderSupplier = inputInMemoryGroupedTopNBuilderSupplier;
```

Comment:
can you add requireNonNull checks for all the non-primitive fields

Reply:
Will do